simplex/storage.go

```diff
@@ -0,0 +1,275 @@
+func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finalization simplex.Finalization) error {
+	bh := block.BlockHeader()
+	if bh.Seq == 0 {
+		s.log.Error("attempted to index genesis block")
+		return errGenesisIndexed
+	}
+
+	currentHeight := s.height.Load()
+	if currentHeight != bh.Seq {
+		s.log.Error("attempted to index block with mismatched sequence number",
+			zap.Uint64("expected", currentHeight),
+			zap.Uint64("got", bh.Seq))
+		return fmt.Errorf("%w: expected %d, got %d", errUnexpectedSeq, currentHeight, bh.Seq)
+	}
+
+	s.blockTracker.lock.Lock()
+	if s.blockTracker.lastIndexed != bh.Prev {
+		s.blockTracker.lock.Unlock()
+		s.log.Error("attempted to index block with mismatched previous digest",
+			zap.Stringer("expected", s.blockTracker.lastIndexed),
+			zap.Stringer("got", bh.Prev))
+
+		return fmt.Errorf("%w: expected %s, got %s", errMismatchedPrevDigest, s.blockTracker.lastIndexed, bh.Prev)
+	}
+	s.blockTracker.lock.Unlock()
+
+	if !bytes.Equal(bh.Digest[:], finalization.Finalization.Digest[:]) {
+		s.log.Error("attempted to index block with mismatched digest",
+			zap.Stringer("expected", bh.Digest),
+			zap.Stringer("got", finalization.Finalization.Digest))
+		return fmt.Errorf("%w: expected %d, got %d", errMismatchedDigest, bh.Digest, finalization.Finalization.Digest)
+	}
+
+	if finalization.QC == nil {
+		s.log.Error("attempted to index block with no quorum certificate", zap.Stringer("blockID", bh.Digest))
+		return errInvalidQC
+	}
+
+	seqBuff := make([]byte, 8)
+	binary.BigEndian.PutUint64(seqBuff, currentHeight)
+
+	finalizationBytes := finalizationToBytes(finalization)
+	if err := s.db.Put(seqBuff, finalizationBytes); err != nil {
+		return fmt.Errorf("failed to store finalization: %w", err)
+	}
+
+	err := s.blockTracker.indexBlock(ctx, bh.Digest)
+	if err != nil {
+		return fmt.Errorf("failed to index block: %w", err)
+	}
+
+	s.height.Add(1) // only increment height after successful indexing
+	return nil
+}
+
+// getGenesisBlock returns the genesis block wrapped as a Block instance.
+func getGenesisBlock(ctx context.Context, config *Config, blockTracker *blockTracker) (*Block, error) {
+	snowmanGenesis, err := getBlockAtHeight(ctx, config.VM, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	genesis := &Block{
+		metadata:     genesisMetadata,
+		blockTracker: blockTracker,
+		vmBlock:      snowmanGenesis,
+	}
+
+	// set the digest
+	bytes, err := genesis.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	genesis.digest = computeDigest(bytes)
+
+	return genesis, nil
+}
+
+// retrieveFinalization retrieves the finalization at [seq].
+// If the finalization is not found, it returns false.
+func (s *Storage) retrieveFinalization(seq uint64) (simplex.Finalization, error) {
+	seqBuff := make([]byte, 8)
+	binary.BigEndian.PutUint64(seqBuff, seq)
+	finalizationBytes, err := s.db.Get(seqBuff)
+	if err != nil {
+		if err == database.ErrNotFound {
+			return simplex.Finalization{}, simplex.ErrBlockNotFound
```

Comment:
Does simplex have special handling for `ErrBlockNotFound`?

I don't have a problem with returning this sentinel error, just mainly interested on if this is something simplex actually uses.

Reply:
```
Does simplex have special handling for ErrBlockNotFound?
```

It has special handling for anything but that - we [put](https://github.com/ava-labs/Simplex/blob/main/epoch.go#L100) the epoch in a state of "I am halted". In the future we can cleverly shut down the chain.